# [\#10828 PR](https://github.com/ethereum/solidity/pull/10828) `merged`: [SMTChecker] Add msgvalue to cex if > 0

#### <img src="https://avatars.githubusercontent.com/u/504195?u=ce2facd14af9fd474ebff49f0d44891f56f7500f&v=4" width="50">[leonardoalt](https://github.com/leonardoalt) opened issue at [2021-01-20 19:51](https://github.com/ethereum/solidity/pull/10828):

This PR adds the missing constraint `msg.value = 0` for nonpayable functions, and shows the actual value in counterexamples if greater than 0.

#### <img src="https://avatars.githubusercontent.com/u/504195?u=ce2facd14af9fd474ebff49f0d44891f56f7500f&v=4" width="50">[leonardoalt](https://github.com/leonardoalt) commented at [2021-01-21 09:53](https://github.com/ethereum/solidity/pull/10828#issuecomment-764516913):

@axic updated with your comments

#### <img src="https://avatars.githubusercontent.com/u/504195?u=ce2facd14af9fd474ebff49f0d44891f56f7500f&v=4" width="50">[leonardoalt](https://github.com/leonardoalt) commented at [2021-01-21 14:29](https://github.com/ethereum/solidity/pull/10828#issuecomment-764679976):

@hrkrshnn we can't add the second test you suggested because using `msg.value` in a non-payable function is a type error.

#### <img src="https://avatars.githubusercontent.com/u/13174375?u=52d702cb6bec53b561afa293cf9cd53ef7a63924&v=4" width="50">[hrkrshnn](https://github.com/hrkrshnn) commented at [2021-01-21 14:34](https://github.com/ethereum/solidity/pull/10828#issuecomment-764685159):

@leonardoalt how about this?

```
contract C {
	bool lock = true;
	function f() public {
	  lock = false;
	  g();
	  lock = true;
	}
	function g() public payable {
		require(lock == false);
		assert(msg.value == 0);
	}
}
``` 

Edit: not sure about this test.

#### <img src="https://avatars.githubusercontent.com/u/504195?u=ce2facd14af9fd474ebff49f0d44891f56f7500f&v=4" width="50">[leonardoalt](https://github.com/leonardoalt) commented at [2021-01-21 14:36](https://github.com/ethereum/solidity/pull/10828#issuecomment-764686611):

Nice, that's a good one

#### <img src="https://avatars.githubusercontent.com/u/504195?u=ce2facd14af9fd474ebff49f0d44891f56f7500f&v=4" width="50">[leonardoalt](https://github.com/leonardoalt) commented at [2021-01-21 14:45](https://github.com/ethereum/solidity/pull/10828#issuecomment-764692787):

There's this one test left to fix:
```
    try_string_literal_to_bytes_array - smtCheckerTests.try_catch

    unexpected exception

    UNCAUGHT EXCEPTION:
    - file: unknown location
    - line: 0


    EXCEPTION STACK TRACE: --------------
    memory access violation at address: 0x00000030: no mapping at fault address
    -------------------------------------
 ```
 It doesn't happen locally so I'm investigating inside CI. Tried to debug it, but doesn't happen in debug mode.

#### <img src="https://avatars.githubusercontent.com/u/504195?u=ce2facd14af9fd474ebff49f0d44891f56f7500f&v=4" width="50">[leonardoalt](https://github.com/leonardoalt) commented at [2021-01-21 19:46](https://github.com/ethereum/solidity/pull/10828#issuecomment-764893528):

@hrkrshnn can you take another look?


-------------------------------------------------------------------------------



[Export of Github issue for [ethereum/solidity](https://github.com/ethereum/solidity). Generated on 2024.12.15 at 06:45:24.]
